eng/CodeAnalysis.src.globalconfig

```diff
@@ -679,6 +688,18 @@ dotnet_diagnostic.CA2260.severity = warning
 # CA2261: Do not use ConfigureAwaitOptions.SuppressThrowing with Task<TResult>
 dotnet_diagnostic.CA2261.severity = warning
 
+# CA2262: Set 'MaxResponseHeadersLength' properly
+dotnet_diagnostic.CA2262.severity = warning
+
+# CA2263: Prefer generic overload when type is known
+dotnet_diagnostic.CA2263.severity = info
```

Comment:
Is there a problem / difficulty with this diagnostic that prevents us from having it be a warning?

Reply:
There is a some cases where the suggested API is calling the type overload internally, I assume we would not like to flag them, let me know if you have other suggestion. There also some that looks could be more efficient and should be fixed, I will update them with this PR, overall around 200 findings with duplicates
[CA2263 findings.txt](https://github.com/dotnet/runtime/files/14504741/CA2263.findings.txt)

Reply:
I think we should go through and prepare the change to fix them, and then evaluate any cases we believe are problematic. Two reasons:
1. One of the reasons we aggressively consume new analyzers into dotnet/runtime is to vet the new analyzers. dotnet/runtime is a large codebase, and is a good first line of defense against problems with an analyzer's implementation or even high-level goal. It'd be good to understand which of those 200 warnings are real things to be fixed and which are noise that might demand changes to the analyzer.
2. We're hopefully only adding analyzers we believe lead to making code better. We want that betterness in dotnet/runtime.

Reply:
(That can/should certainly be separate from this PR.)

Reply:
> I think we should go through and prepare the change to fix them, and then evaluate any cases we believe are problematic. Two reasons:

Sure, I did evaluate them and concluded that those are not issue with the analyzer and also not worth fixing in the runtime, here is the reasons, please let me know if you have any comment/suggestion:
Excluded these cases from fixing:
1. Excluded fixing the cases when the project builds for netstandard or .NET framework where the generic overload doesn't exit
2.  Cases where the generic overload implementation just calling the `Type` overload, like: `public T CreateDelegate<T>() where T : Delegate => (T)CreateDelegate(typeof(T));`. I think there is nothing to fix in the analyzer, unless we want to check the implementation, which I did not think worth a fix, because the analyzer suggestion is more about AOT compliance, external applications should change to use the generic overload if they care about AOT or similar tooling
3. The generic overload implementation using typeof(T) internally (this is kind of similar to the case 2 but the implementation not exactly calling the `Type` overload). For example: `Marshal.SizeOf<T>()`, `Marshal.PtrToStructure<T>(nint)` etc. By my evaluation these doesn't look worth fixing, but not that sure with my evaluation. No fix needed in the analyzer with the same reason as above, from AOT perspective this is not false positive. 

Here are the all cases that I did not fix: [CA2263 findings_NotFixed.txt](https://github.com/dotnet/runtime/files/14513603/CA2263.findings_NotFixed.txt)

> (That can/should certainly be separate from this PR.) 

I have fixed the ones that worth fixing, PTAL

Reply:
>The generic overload implementation using typeof(T) internally (this is kind of similar to the case 2 but the implementation not exactly calling the Type overload). For example: Marshal.SizeOf<T>()

`Marshal.SizeOf<T>()` does not delegate to `Marshal.SizeOf(Type)`. Why aren't these worth fixing?

> Cases where the generic overload implementation just calling the Type overload

I'm still wondering why we'd want to warn customers about these cases but then not enable the rule ourselves because it's warning about these cases and we don't want to fix them. If we consider these noise, why wouldn't others?

Reply:
Here is my thought process: the main reason of adding [Prefer generic overload when type is known](https://github.com/dotnet/runtime/issues/68243) analyzer was for tooling like AOT or trimming. If the generic overload is just calling the `Type` overload changing the call to use generic overload is useful only for that purpose. For runtime I assumed we already have done AOT trim analysis for all code base and those calls are kept as is it is safe or worked around already. So, changing them to use generic overload could only cause perf degradation. Though sI am might be wrong, please let me know.

> Marshal.SizeOf<T>() does not delegate to Marshal.SizeOf(Type). Why aren't these worth fixing?

Was not sure if it is worth fixing, the generic overload might less performant, if you are suggesting to fix them I can do that. 

Also, if you are suggesting setting the severity to `warning` and/or make adjustment to the analyzer let me know, I just did what felt right and I might have mistaken for some aspect.

Reply:
@jkotas, my preference would be to enable the analyzer and follow its recommendations in these cases (or else revise the analyzer if we don't believe in it), but I know you had some objections to adding such CreateDelegate generic overloads initially. Do you have any concerns about switching these over?

Reply:
> my preference would be to enable the analyzer and follow its recommendations in these cases (or else revise the analyzer if we don't believe in it)

I will handle that with a separate PR, we can continue the discussion there

Reply:
> @jkotas, my preference would be to enable the analyzer and follow its recommendations in these cases (or else revise the analyzer if we don't believe in it), but I know you had some objections to adding such CreateDelegate generic overloads initially. Do you have any concerns about switching these over?

I see most of these analyzer suggestions as syntax sugar improvements. The code is a bit less verbose, but the type loader and the JIT have to do a bit more work. I do not have concerns about these changes in general.

The only ones where I have my usual concern are places like the System.Linq.Expression ones where this syntax sugar improvement is applied in a frequently instantiated generic and the extra runtime work gets multiplied by a large number. It is what I have commented on before that you probably remember: https://github.com/dotnet/runtime/pull/36680#discussion_r428821774 . We should #pragma disable for these few places if we want to enable this analyzer globally.

I have done a cursory look over https://github.com/dotnet/runtime/files/14513603/CA2263.findings_NotFixed.txt . Majority places are either in libraries that target downlevel platforms where the syntax sugar improvement would have to be under an ifdef and thus it is not worth it; and in `System.DirectoryServices` and `System.Data.OleDb` libraries that we try to avoid touching (these libraries did not get interop modernization pass, AOT compatibility treatment, ...).

> For runtime I assumed we already have done AOT trim analysis for all code base and those calls are kept as is it is safe or worked around already.

Right, the AOT concerns are checked by the AOT/trimming analyzer, so this analyzer is not going to warn about any additional problematic places over what AOT/trimming analyzer warns about already.

Reply:
> Right, the AOT concerns are checked by the AOT/trimming analyzer, so this analyzer is not going to warn about any additional problematic places over what AOT/trimming analyzer warns about already.

The AOT rules are also more fine grained than just "generic overload good, non-generic overload bad". Using the non-generic overload with a `typeof` that is statically visible is often okay because it can be analyzed same way as if it was a generic call (it does not produce a trimming/AOT warning in the respective analyzers).

The AOT/trimming analyzers are in the best position to know what is and isn't safe for AOT/trimming. This analyzer should probably not concern itself with what is or isn't trim/AOT safe.

Reply:
> The AOT/trimming analyzers are in the best position to know what is and isn't safe for AOT/trimming. This analyzer should probably not concern itself with what is or isn't trim/AOT safe.

That was the initial purpose of this analyzer, the difference from AOT/trimming analyzers are [listed in the issue](https://github.com/dotnet/runtime/issues/68243#issuecomment-1105706974) 

>  - Will generate a warning on reference + it has to have a generic safe overload
>  - The codefixer would act on this new set of rules and suggest a different overload beside the other two options
>  - The analyzer will be enabled by default and not restained to EnableAotAnalyzer

Though not sure if the AOT/trimming analyzers are now changed and made these statements false.

Please let me know if you have any improvement suggestions for this analyzer, or if you think the analyzer is useless now

Reply:
> > my preference would be to enable the analyzer and follow its recommendations in these cases (or else revise the analyzer if we don't believe in it)
> 
> I will handle that with a separate PR, we can continue the discussion there

> I see most of these analyzer suggestions as syntax sugar improvements. The code is a bit less verbose, but the type loader and the JIT have to do a bit more work. I do not have concerns about these changes in general.
> 
> The only ones where I have my usual concern are places like the System.Linq.Expression ones where this syntax sugar improvement is applied in a frequently instantiated generic and the extra runtime work gets multiplied by a large number. It is what I have commented on before that you probably remember: https://github.com/dotnet/runtime/pull/36680#discussion_r428821774 . We should #pragma disable for these few places if we want to enable this analyzer globally.
> 
> I have done a cursory look over https://github.com/dotnet/runtime/files/14513603/CA2263.findings_NotFixed.txt . Majority places are either in libraries that target downlevel platforms where the syntax sugar improvement would have to be under an ifdef and thus it is not worth it; and in `System.DirectoryServices` and `System.Data.OleDb` libraries that we try to avoid touching (these libraries did not get interop modernization pass, AOT compatibility treatment, ...).

Sounds like we would like to keep the analyzer severity info level, apply the code-fix for all cases excluding projects that target downlevel platforms plus also excluding `System.Linq.Expression`, `System.DirectoryServices` and `System.Data.OleDb` projects.

Reply:
> Please let me know if you have any improvement suggestions for this analyzer, or if you think the analyzer is useless now

This analyzer is first and foremost a code style analyzer/fixer. I do not think it is useless, but I am not sure whether it should be enabled by default. Do we have a prior art for analyzers like that?

Reply:
> This analyzer is first and foremost a code style analyzer/fixer.

If that's really the case, we shouldn't have it at all; we're generally not in the business of including style analyzers in this library.

But it's that really the case?

Reply:
> This analyzer is first and foremost a code style analyzer/fixer. I do not think it is useless, but I am not sure whether it should be enabled by default. Do we have a prior art for analyzers like that?

Then this rule should start with "IDE" prefix https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/categories#style-rules. I think "IDE" rules should not added in `roslyn-analyzers` repo, probably in `roslyn`?

Reply:
In some cases, the fix suggested by the analyzer may result in a minor performance improvement. Example: Replacing `Enum.GetName` with `EnumGetName<TEnum>` will avoid boxing. I am not aware of anything else.

Reply:
> In some cases, the fix suggested by the analyzer may result in a minor performance improvement. Example: Replacing `Enum.GetName` with `EnumGetName<TEnum>` will avoid boxing. I am not aware of anything else.

Yes, and that is why I fixed those cases with this PR, thought the cases where the generic overload implementation just calling the `Type` overload, will have some perf degradation, for other case like `Marshal.SizeOf<T>()` and  `Marshal.SizeOf(Type)` the perf impact was not obvious, so did not fix. Anyway, I don't think with the analyzer we could evaluate the perf impact.

Based on your and @stephentoub's feedback I put up a [PR](https://github.com/dotnet/runtime/pull/100490) that fixes all cases excluding the projects that target down-level platforms and `System.Linq.Expression`.

Reply:
> > This analyzer is first and foremost a code style analyzer/fixer.
> 
> If that's really the case, we shouldn't have it at all; we're generally not in the business of including style analyzers in this library.
> 
> But is that really the case?

I think that is the case, so what we should do with the analyzer now?

Reply:
> I think that is the case, so what we should do with the analyzer now?

The fact that it catches things like changing Enum calls from non-generic to generic makes me still excited about it. I was just fixing up an internal codebase and found it useful specifically for that.